### PR TITLE
Changed Pearson r to handle masked data in a more reasonable way

### PR DESCRIPTION
**Pull request by niallrobinson**
Peaerson r now only considers a subset of the data where corresponding data in both cubes are not masked. This has implications for the assumed length of the cubes, and can change the calculated r values

in response to https://github.com/SciTools/iris/issues/1534

### CODE REVIEW

File: lib/iris/analysis/stats.py [hunk 1]
```diff
@@ -30,7 +30,8 @@ def _get_calc_view(cube_a, cube_b, corr_coords):
     """
     This function takes two cubes and returns cubes which are
     flattened so that efficient comparisons can be performed
-    between the two.
+    between the two. If the arrays are maksed then only values
```
Comment (marqh) [inline lib/iris/analysis/stats.py]: just a typo *masked

File: lib/iris/analysis/stats.py [hunk 2]
```diff
@@ -94,7 +95,29 @@ def _get_calc_view(cube_a, cube_b, corr_coords):
...
+    else:
+        b_not_masked = True
+
+    both_not_masked = a_not_masked & b_not_masked
+    try:
+        # compress to good values using mask array
+        return_a = reshaped_a.compress(both_not_masked)
+        return_b = reshaped_b.compress(both_not_masked)
+    except ValueError:
+        # expect when masks are just non-array True/False
```
Comment (marqh) [inline lib/iris/analysis/stats.py]: please could you expand the comment a little to explain why any ValueError is acceptable to be excepted, I feel I am missing a bit of logic

File: lib/iris/analysis/stats.py [hunk 3]
```diff
@@ -94,7 +95,29 @@ def _get_calc_view(cube_a, cube_b, corr_coords):
     reshaped_b = data_b.transpose(slice_ind+res_ind)\
                        .reshape(dim_i_len, dim_j_len)
 
-    return reshaped_a, reshaped_b, res_ind
+    # Remove data where one or both cubes are masked
+    # First deal with the case that either cube is unmasked
+    # Collapse masks to the dimension we are correlating over (0th)
+    if np.ma.is_masked(reshaped_a):
```
Comment (ajdawson) [inline lib/iris/analysis/stats.py]: Should be just `ma.is_masked(reshaped_a)` with an additional import at the top of the file `import numpy.ma as ma`. This will match the style used elsewhere.

File: lib/iris/analysis/stats.py [hunk 4]
```diff
@@ -94,7 +95,29 @@ def _get_calc_view(cube_a, cube_b, corr_coords):
...
+        a_not_masked = np.logical_not(reshaped_a.mask).any(axis=1)
+    else:
+        a_not_masked = True
+    if np.ma.is_masked(reshaped_b):
```
Comment (ajdawson) [inline lib/iris/analysis/stats.py]: As above.

**Comment by ajdawson** [comment]
I found the logic a bit hard to follow on first read. I'd personally prefer something a bit more concise, I think the following is equivalent to lines 101-120:

```python
a_mask = ma.is_masked(reshaped_a) and reshaped_a.mask
b_mask = ma.is_masked(reshaped_b) and reshaped_b.mask
mask_union = a_mask | b_mask
if not np.isscalar(mask_union):
    return_a = reshaped_a.compress(~mask_union)
    return_b = reshaped_b.compress(~mask_union)
else:
    return_a = reshaped_a
    return_b = reshaped_b
return return_a, return_b, res_ind
```

I guess there could be arguments against this too, but at least there are fewer lines of code to understand.

**Comment by ajdawson** [comment]
@niallrobinson - This work would be superseded by the changes in #1714. The changes there make handling masks and weights a  more manageable task.

**Comment by ajdawson** [comment]
I'm closing this since the whole pearsonr code has subsequently been re-written.
